README.md
# ecslog

`ecslog` is a CLI for pretty-printing and filtering log files in
[ecs-logging](https://www.elastic.co/guide/en/ecs-logging/overview/master/intro.html)
format. It is a single (Go-implemented) binary, with builds for Linux, Mac, and
Windows. Its default styling should play nice with both light and
dark-background terminals.

![ecslog demo screenshot](./docs/img/demo-light-and-dark.png)

# Install

- For homebrew users:

        brew install trentm/tap/ecslog
        # OR 'brew upgrade trentm/tap/ecslog' if you have it already

- Or download a pre-built binary package from [the releases page](https://github.com/trentm/go-ecslog/releases)
  and copy to somewhere on your PATH.

- Or you can build from source via:

        git clone git@github.com:trentm/go-ecslog.git
        cd go-ecslog
        make  # produces "./ecslog", a single binary you can put on your PATH
        ./ecslog --version

Then, try it on a demo log file:

    curl -s https://raw.githubusercontent.com/trentm/go-ecslog/main/demo.log \
        | ecslog

# Introduction

`ecslog` pretty-prints log data in the given file argument(s):

    ecslog /var/log/some.log [other-log-files...]

or passed in on stdin

    node server.js | ecslog
    tail -f /var/log/some.log | ecslog
    docker logs -f my-container | ecslog

The default behaviour can be customised by CLI options (use `ecslog --help` to
list all options) and/or a `~/.ecslog.toml` config file (see
[Configuration](#configuration) below).  The rest of this section describes some
features of ecslog.


## Level filtering

Use `-l LEVEL-NAME` to only show log records at that level or above.

![ecslog level filtering](./docs/img/level-filtering.png)

([ECS](https://www.elastic.co/guide/en/ecs/current/ecs-log.html#field-log-level)
does not mandate level names, so `ecslog` uses [a best-effort ordering](https://github.com/trentm/go-ecslog/blob/v0.4.0/internal/ecslog/ecslog.go#L150-L168)
of level names in common usage.)


## KQL filtering

Use `-k KQL` to filter log records using [KQL](https://www.elastic.co/guide/en/kibana/current/kuery-query.html), as you would in Kibana.

![ecslog KQL filtering](./docs/img/kql-filtering.png)

Some examples:

- Filter on the "message" field by default: `ecslog ./examples/apm-server.log -k '*ILM*'`
- Filter on logger name (the `log.logger` field in ECS): `ecslog ./examples/apm-server.log -k 'log.logger: pipeline'`
- Filter on slow requests: `ecslog ./example/apm-server.log -k 'event.duration > 500000'`

Note that this is a subset of KQL and necessarily slightly adapted for use on log files without an Elasticsearch mapping for field types. See [internal/kqlog/README.md](./internal/kqlog/README.md) for details.


## Include/exclude fields from rendering

Sometimes it can help to focus by eliding some distracting fields. Use `-x FIELD,FIELD,...`
to exclude named fields from the rendered output.

![ecslog exclude fields](./docs/img/exclude-fields.png)

Or use `-i FIELD,FIELD,...` to exclude all extra fields (those after the title
line) **except** those given.


## `@timestamp` diff highlighting

By default, `ecslog` will highlight the change in a log record's `@timestamp`
from the previous log record. With the "default" formatter, the changed part
of the timestamp is underlined. For example:

![ecslog @timestamp diff highlighting](./docs/img/timestamp-diff-highlighting.png)

This can be turned off with the [`timestampShowDiff=false` config var](#config-timestampshowdiff).


## `ecsLenient` for almost-ecs-logging format logs

The [ecs-logging spec](https://github.com/elastic/ecs-logging/blob/master/spec/spec.json)
mandates all of the `ecs.version`, `log.level`, and `@timestamp` fields. Given
that ecs-logging is a relatively new spec, some applications, in the process of
migrating, are *almost* conformant. Add `ecsLenient=true` to `~/.ecslog.toml` to
only require that log lines have *one* of these fields to be considered an
ecs-logging record.

<!--
- elasticsearch: https://github.com/elastic/elasticsearch/issues/73256
- kibana: Missing log.level (because hapi "tags"). I haven't opened an issue.
- apm-server: https://github.com/elastic/beats/issues/25922
-->


## Output formats

`ecslog` has multiple output formats for rendering ECS logs that may be selected
via the `-f, --format NAME` option. Some formats are *lossy*, i.e. do not render
all fields, typically for compactness. Formats labelled as "lossless" have one
exception: the "ecs.version" field is typically not rendered.

- `default`: A default lossless that format renders each log record with a
  "title line" -- which includes core and common fields -- followed by all
  remaining extra fields. Roughly:

  ```
  [@timestamp] LOG.LEVEL (log.logger/service.name on host.hostname): message
      extraKey1: extraValue1-as-multiline-jsonish
      extraKey2: extraValue2-as-multiline-jsonish
  ```

  where "multiline jsonish" means 4-space-indented JSON with the one special
  case that multiline string values are printed indented and with newlines.
  For example, `error.stack_trace` in the following:

  ```
  [2021-02-11T06:24:53.251Z]  WARN (myapi on purple.local): something went wrong
      error: {
          "type": "Error"
          "message": "boom"
          "stack_trace":
              Error: boom
                  at pino/examples/express-simple.js:67:15
                  at Layer.handle [as handle_request] (pino/node_modules/express/lib/router/layer.js:95:5)
                  ...
  ```

  The "key: value" format, for simple values, has the benefit of being usable
  as-is as a KQL filter with the `-k KQL` option.  The format of the title line
  may change in future versions.

- `compact`: A lossless format similar to "default", but attempts are made
  to make the "extraKey" info more compact by balancing multiline JSON with
  80-column output.

- `ecs`: The native/raw ECS format, ndjson.

- `simple`: A *lossy* format that simply renders `LOG.LEVEL: message`. If extra
  fields (other than the core "@timestamp" and "ecs.version" fields) are being
  elided, a ellipsis is appended to the line.


## `--strict` to filter out non-ecs-logging lines

By default `ecslog` will pass through non-ecs-logging lines unchanged, which is
useful for mixed-format logs that are typically in the real world -- for example,
structured logging mixed with debug printfs. However, sometimes it can be
useful to limit to just ecs-logging lines. Use the `--strict` option for this.

One use case for this is in a pipeline that will process JSON logs, say, with
`jq`. For example, this will filter an example "apm-server" log file to
records with the 'event.duration' field (`-k 'event.duration: *'`), exclude
non-ecs-logging records (`--strict`), format as raw JSON (`-f ecs`) and then
pipe to `jq` to extract the single JSON field:

```shell
$ ecslog examples/apm-server.log -k 'event.duration: *' --strict -f ecs | jq '."event.duration"'
166823
104904
76471
90210
149853
10226197680
...
```


# Configuration

Any of the following `ecslog` options can be set in a `~/.ecslog.toml` file.
See https://toml.io/ for TOML syntax information.  The `--no-config` option can
be used to ignore `~/.ecslog.toml`, if there is one.

An example config:

```toml
format="compact"
maxLineLen=32768
ecsLenient=true
```

### config: format

Set the output format name (a string, equivalent of `-f, --format` option).
Valid values are: "default" (the default), "compact", "ecs", "simple".

```toml
format="default"
```

### config: color

A color mode string for whether output should be colorized.  Valid values are:
"auto" (the default), "yes", "no". "auto" will colorize if the output stream
is a TTY.

```toml
color="auto"
```

### config: maxLineLen

Set the maximum number of bytes long for a single line that will be considered
for processing. Longer lines will be treated as if they are not ecs-logging
records.  Valid values are: -1 (to use the default 16384), or a value between 1
and 1048576 (inclusive).

```toml
maxLineLen=16384
```

### config: ecsLenient

Some JSON logs are "ECS compatible" in that they attempt to follow [ECS general
guidelines](https://www.elastic.co/guide/en/ecs/current/ecs-guidelines.html) --
have a `@timestamp`, use the specified data types for ECS fields, set
`ecs.version` -- but are not "ecs-logging compliant" because they are missing
one or more of `@timestamp`, `ecs.version`, or `log.level`.

By default `ecslog` will skip rendering for any log line that does not have
those three fields. Set `ecsLenient` to true to tell `ecslog` to attempt to
rendering any log record that has **at least one** of these fields.

```toml
ecsLenient=false
```

### config: timestampShowDiff

If coloring the output (see [config: color](#config-color) above), by default
`ecslog` will style the change in the timestamp from the preceding log record.
Set this config var to `false` to turn off this styling.

```toml
timestampShowDiff=true
```


# Bugs

If you find a crash or some other issue with `ecslog`, please
[create an issue](https://github.com/trentm/go-ecslog/issues/new).

The `ECSLOG_DEBUG` environment variable can be set to get some internal limited
debugging information on stderr. For example:

    ECSLOG_DEBUG=1 ecslog ...

Internal debug logging is disabled if `ECSLOG_DEBUG` is unset, or is set
to one of: the empty string, "0", or "false".


CHANGELOG.md
# ecslog changelog

## v0.6.0

- Make rendering of empty arrays and objects in the extra fields more compact.
  They are now rendered as "[]" and "{}" rather than "[\n]" and "{\n}".
  ([#27](https://github.com/trentm/go-ecslog/pull/27))

## v0.5.0

- Add a `-i, --include-fields FIELDS` flag for including only certain fields.
  When given, *only* those extra fields are rendered in the output. Note that
  `-i FIELDS` do not apply to the title line fields. While this sounds
  incongruous with `-x FIELDS`, which *does* apply to title line fields, I
  think it makes for more natural usage.
  (by @jalvz, [#15](https://github.com/trentm/go-ecslog/pull/15))

- Fix a bug where `format="..."` from the config file was not being read.

- Add "DEPRECATION" level (with same sort order as WARN) found in elasticsearch
  8.x logs.

- Fix an issue where coloring did not handle uppercase level names.

## v0.4.0

- Add [`@timestamp` diff highlighting](README.md#timestamp-diff-highlighting):
  the part of the timestamp that has changed from the preceding record is
  underlined (in the default color scheme). This highlighting can be turned
  off with the `timestampShowDiff: false` config var.
  ([#20](https://github.com/trentm/go-ecslog/pull/20))

- Add `ecsLenient: false` config option to allow rendering of lines that are
  likely ECS-compatible, but do not have all three required ecs-logging fields:
  `@timestamp`, `ecs.version`, `log.level`. Only one of those three is required
  to be rendered.

  This intentially doesn't have a command-line option for now.  Currently it is
  considered a crutch for ES 8.x and Kibana 8.x logs that, at time of writing,
  are missing one or two of the above fields. If that is long-standing,
  `ecsLenient: true` might eventually become the default.

## v0.3.0

- Use goreleaser for releases. The "Version" generally includes the leading "v"
  now. Built binaries should be reproducible from a given commit. They should
  be smaller now ("-s -w" in ldflags). Homebrew support.

- Add `-x, --exclude-fields ...` option to exclude fields from the rendering.
  For example, say you have log records that always has static "foo" and
  "bar" fields. They add two lines to the output for every record, wasting
  space.  `ecslog -x foo,bar` will remove them from the record before rendering
  it, helping with info density.

- Fix a bug where `-f FORMAT` would be ignored if there was a "format: FORMAT"
  in "~/.ecslog.toml".

## v0.2.0

- Add `--strict` option that will suppress input lines that are not valid
  ecs-logging records. Normally non-ecs-logging records are passed through
  unchanged.

- Support there not being a "message" field (allowed in ecs-logging spec
  in https://github.com/elastic/ecs-logging/pull/55).

- Fix a bug in the "simple" formatter, where the ellipsis would always be
  printed because the "@timestamp" field was not discounted.

- Refactor the read loop to handle very long lines without crashing, and without
  using unbounded memory. One side-effect -- due to the usage of
  `bufio.Reader.ReadLine` -- is that ecslog output will always finish with a
  newline, even if the input did not.

- Potentially much faster passing through unprocessed lines, moving to
  `out.Write` instead of unnecessary usage of `fmt.Fprintln`.

## v0.1.0

- Added KQL filtering via `ecslog -k,--kql KQL-FILTER`. For example:

        cat demo.log | ./ecslog -k error:*
        cat demo.log | ./ecslog -k 'http.response.status_code>=200'

  See [the Kibana KQL docs](https://www.elastic.co/guide/en/kibana/current/kuery-query.html)
  for an introduction to KQL, and see [the kqlog README](./internal/kqlog/README.md)
  for notes on the subset of KQL implemented.


internal/ecslog/ecslog.go
package ecslog

// Shared stuff for `ecslog` that isn't specific to the CLI.

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"

	"github.com/mattn/go-isatty"
	"github.com/trentm/go-ecslog/internal/ansipainter"
	"github.com/trentm/go-ecslog/internal/jsonutils"
	"github.com/trentm/go-ecslog/internal/kqlog"
	"github.com/trentm/go-ecslog/internal/lg"
	"github.com/valyala/fastjson"
)

// Version is the semver version of this tool.
const Version = "v0.6.0"

const defaultMaxLineLen = 16384

// Renderer is the class used to drive ECS log rendering (aka pretty printing).
type Renderer struct {
	parser            fastjson.Parser
	painter           *ansipainter.ANSIPainter
	formatName        string
	formatter         Formatter
	maxLineLen        int
	excludeFields     []string
	includeFields     []string
	ecsLenient        bool
	timestampShowDiff bool
	levelFilter       string
	kqlFilter         *kqlog.Filter
	strict            bool

	line             []byte // the raw input line
	logLevel         string // cached "log.level", read during isECSLoggingRecord
	lastTimestampBuf []byte // buffer to hold lastTimestamp values
	lastTimestamp    []byte // last @timestamp (a slice of lastTimestampBuf)
}

// NewRenderer returns a new ECS logging log renderer.
//
// - `shouldColorize` is one of "auto" (meaning colorize if the output stream
//   is a TTY), "yes", or "no"
// - `colorScheme` is the name of one of the colors schemes in
//   ansipainter.PainterFromName
// - `maxLineLen` a maximum number of bytes for a line that will be considered
//   for log record processing. It must be a positive number between 1 and
//   1048576 (2^20), or -1 to use the default value (16384).
// - `ecsLenient` is a bool indicating if rendering should be lenient in
//   validating if a JSON line is an ecs-logging record. Strictly (the default)
//   to be an ecs-logging record it must have ecs.version, log.level, and
//   @timestamp. If this option is true, it will only require *one* of those
//   fields to exist.
// - `timestampShowDiff` is a bool indicating if the @timestamp diff from the
//   preceding log record should be styled.
func NewRenderer(shouldColorize, colorScheme, formatName string, maxLineLen int, excludeFields, includeFields []string, ecsLenient, timestampShowDiff bool) (*Renderer, error) {
	// Get appropriate "painter" for terminal coloring.
	var painter *ansipainter.ANSIPainter
	if shouldColorize == "auto" {
		if isatty.IsTerminal(os.Stdout.Fd()) {
			shouldColorize = "yes"
		} else {
			shouldColorize = "no"
		}
	}
	switch shouldColorize {
	case "yes":
		var ok bool
		painter, ok = ansipainter.PainterFromName[colorScheme]
		if !ok {
			var known []string
			for n := range ansipainter.PainterFromName {
				known = append(known, n)
			}
			sort.Strings(known)
			return nil, fmt.Errorf("unknown color scheme '%s' (known schemes: %s)",
				colorScheme, strings.Join(known, ", "))
		}
	case "no":
		painter = ansipainter.NoColorPainter
		// No point in calculating timestamp diff if not styling.
		timestampShowDiff = false
	default:
		return nil, fmt.Errorf("invalid value for shouldColorize: %s", shouldColorize)
	}

	formatter, ok := formatterFromName[formatName]
	if !ok {
		var known []string
		for n := range formatterFromName {
			known = append(known, n)
		}
		sort.Strings(known)
		return nil, fmt.Errorf("unknown format '%s' (known formats: %s)",
			formatName, strings.Join(known, ", "))
	}

	if maxLineLen == -1 {
		maxLineLen = defaultMaxLineLen
	} else if maxLineLen <= 0 || maxLineLen > 1048576 {
		return nil, fmt.Errorf("invalid maxLineLen, must be -1 or between 1 and 1048576: %d",
			maxLineLen)
	}

	lg.Printf("create renderer: formatName=%q, shouldColorize=%q, colorScheme=%q, maxLineLen=%d\n",
		formatName, shouldColorize, colorScheme, maxLineLen)
	return &Renderer{
		painter:           painter,
		formatName:        formatName,
		formatter:         formatter,
		maxLineLen:        maxLineLen,
		excludeFields:     excludeFields,
		includeFields:     includeFields,
		ecsLenient:        ecsLenient,
		timestampShowDiff: timestampShowDiff,

		// Can a timestamp ever reasonably be longer than 64 chars?
		// "2021-04-15T04:22:29.507Z" is 24.
		lastTimestampBuf: make([]byte, 64),
	}, nil
}

// SetLevelFilter sets the level at which `log.level` filtering is done.
func (r *Renderer) SetLevelFilter(level string) {
	if level != "" {
		r.levelFilter = level
	}
}

// SetKQLFilter sets the KQL statement used for log record filtering.
func (r *Renderer) SetKQLFilter(kql string) error {
	var err error
	if kql != "" {
		r.kqlFilter, err = kqlog.NewFilter(kql, LogLevelLess)
	}
	return err
}

// SetStrictFilter tells the renderer whether to strictly suppress input lines
// that are not valid ecs-logging records.
func (r *Renderer) SetStrictFilter(strict bool) {
	r.strict = strict
}

// levelValFromName is a best-effort ordering of levels in common usage in
// logging frameworks that might be used in ECS format. See `LogLevelLess`
// below. (The actual int values are only used internally and can change between
// versions.)
//
// - zap: https://pkg.go.dev/go.uber.org/zap/#AtomicLevel.MarshalText
// - bunyan: https://github.com/trentm/node-bunyan/tree/master/#levels
// - elasticsearch emits "DEPRECATION" as a level; we equate that with WARN
// - ...
var levelValFromName = map[string]int{
	"trace":       10,
	"debug":       20,
	"info":        30,
	"deprecation": 40,
	"warn":        40,
	"warning":     40,
	"error":       50,
	"dpanic":      60,
	"panic":       70,
	"fatal":       80,
}

// LogLevelLess returns true iff level1 is less than level2.
//
// Because ECS doesn't mandate a set of log level names for the "log.level"
// field, nor any specific ordering of those log levels, this is a best
// effort based on names and ordering from common logging frameworks.
// If a level name is unknown, this returns false. Level names are considered
// case-insensitive.
func LogLevelLess(level1, level2 string) bool {
	val1, ok := levelValFromName[strings.ToLower(level1)]
	if !ok {
		return false
	}
	val2, ok := levelValFromName[strings.ToLower(level2)]
	if !ok {
		return false
	}
	return val1 < val2
}

// LevelNameOrderStr returns a string with all the known level names in order
// in the format "trace < debug < ...".
func LevelNameOrderStr() string {
	type nv struct {
		name string
		val  int
	}

	var sorted []nv
	for n, v := range levelValFromName {
		sorted = append(sorted, nv{n, v})
	}

	sort.SliceStable(sorted, func(i, j int) bool {
		return sorted[i].val < sorted[j].val
	})

	var names []string
	var currVal int
	for _, nv := range sorted {
		if nv.val == currVal && len(names) > 0 {
			names[len(names)-1] = names[len(names)-1] + "/" + nv.name
		} else {
			names = append(names, nv.name)
			currVal = nv.val
		}
	}
	return strings.Join(names, " < ")
}

// isECSLoggingRecord returns true iff the given `rec` has the required
// ecs-logging fields: @timestamp, ecs.version, and log.level (all
// strings). If `message` is present, it must be a string.
//
// Caveat: If `r.ecsLenient` then only one of the three "required" fields is
// needed to be considered an "ecs-logging" record.
//
// Side-effect: r.logLevel is cached on the Renderer for subsequent use.
func (r *Renderer) isECSLoggingRecord(rec *fastjson.Value) bool {
	logLevel := jsonutils.LookupValue(rec, "log", "level")
	if logLevel != nil && logLevel.Type() == fastjson.TypeString {
		r.logLevel = string(logLevel.GetStringBytes())
		if r.ecsLenient {
			return true
		}
	} else if !r.ecsLenient {
		return false
	}

	timestamp := rec.GetStringBytes("@timestamp")
	if timestamp != nil {
		if r.ecsLenient {
			return true
		}
	} else if !r.ecsLenient {
		return false
	}

	ecsVersion := jsonutils.LookupValue(rec, "ecs", "version")
	if ecsVersion != nil && ecsVersion.Type() == fastjson.TypeString {
		if r.ecsLenient {
			return true
		}
	} else if !r.ecsLenient {
		return false
	}

	message := rec.Get("message")
	if message != nil && message.Type() != fastjson.TypeString {
		// If there is a "message" it must be a string.
		return false
	}

	if r.ecsLenient {
		return false
	}
	return true
}

// RenderFile renders log records from the given open file stream to the given
// output stream (typically os.Stdout).
func (r *Renderer) RenderFile(in io.Reader, out io.Writer) error {
	var b strings.Builder
	eol := []byte{'\n'}

	// For speed we want each processed line to fit in a single buffer that
	// we don't need to copy/extend. That means at least:
	//   maxLineLen + 2 (for '\r\n' line end)
	// However if maxLineLen is configured to something really small, then
	// that could hurt perf, so set a min of 64k (bufio.Scanner's default)
	const minBufSize = 65536
	bufSize := r.maxLineLen + 2
	if bufSize < minBufSize {
		bufSize = minBufSize
	}
	reader := bufio.NewReaderSize(in, bufSize)

	var wasPrefix bool
	for {
		// We use reader.ReadLine() to avoid consuming unbounded memory for
		// a crazy-long line. If a line is longer than our read buffer, then
		// we incrementally write it through unprocessed.
		// `reader.ReadBytes('\n')` uses mem to the size of the input line.
		//
		// Note that due to this from `reader.ReadLine()`
		//    > No indication or error is given if the input ends without
		//    > a final line end.
		// ecslog will always end its output with a newline, even if the input
		// doesn't have one.
		line, isPrefix, err := reader.ReadLine()
		if err == io.EOF {
			return nil
		} else if err != nil {
			return err
		}
		if wasPrefix || isPrefix {
			// This is a line > maxLineLen, so we just want to print it
			// unchanged. The current line continues until `isPrefix == false`.
			wasPrefix = isPrefix
			if !r.strict {
				out.Write(line)
				if !isPrefix {
					out.Write(eol)
				}
			}
			continue
		}

		// For now, do *not* support lines with leading whitespace. Happy to
		// reconsider if there is a real use case.
		if len(line) == 0 || len(line) > r.maxLineLen || line[0] != '{' {
			if !r.strict {
				out.Write(line)
				out.Write(eol)
			}
			continue
		}

		rec, err := r.parser.ParseBytes(line)
		if err != nil {
			lg.Printf("line parse error: %s\n", err)
			if !r.strict {
				out.Write(line)
				out.Write(eol)
			}
			continue
		}

		if !r.isECSLoggingRecord(rec) {
			if !r.strict {
				out.Write(line)
				out.Write(eol)
			}
			continue
		}
		r.line = line

		// `--level info` will drop any log records less than log.level=info.
		if r.levelFilter != "" && LogLevelLess(r.logLevel, r.levelFilter) {
			continue
		}

		if r.kqlFilter != nil && !r.kqlFilter.Match(rec) {
			continue
		}

		for _, xf := range r.excludeFields {
			if len(xf) == 0 {
				continue
			} else if xf == "log.level" {
				// Special case: log.level is already removed and cached on
				// the Renderer.
				r.logLevel = ""
			} else {
				jsonutils.ExtractValue(rec, strings.Split(xf, ".")...)
			}
		}

		r.formatter.formatRecord(r, rec, &b)
		out.Write([]byte(b.String()))
		out.Write(eol)
		b.Reset()
	}
}
